crates/matrix-sdk/tests/integration/encryption/verification.rs

```diff
@@ -99,33 +185,91 @@ impl MockedServer {
                 assert!(params.self_signing_key.is_some());
                 assert!(params.user_signing_key.is_some());
 
-                // We don't do anything particular with the keys, at this point.
+                let mut keys = k.lock().unwrap();
+
+                if let Some(key) = params.master_key {
+                    let deserialized = key.deserialize().unwrap();
+                    let user_id = deserialized.user_id;
+                    keys.master.insert(user_id, key);
+                }
+
+                if let Some(key) = params.self_signing_key {
+                    let deserialized = key.deserialize().unwrap();
+                    let user_id = deserialized.user_id;
+                    keys.self_signing.insert(user_id, key);
+                }
+
+                if let Some(key) = params.user_signing_key {
+                    let deserialized = key.deserialize().unwrap();
+                    let user_id = deserialized.user_id;
+                    keys.user_signing.insert(user_id, key);
+                }
 
                 ResponseTemplate::new(200).set_body_json(json!({}))
             })
             .mount(&server)
             .await;
 
-        let known_devices_clone = known_devices.clone();
-        let device_keys_clone = device_keys.clone();
+        let k = keys.clone();
         Mock::given(method("POST"))
             .and(path("/_matrix/client/unstable/keys/signatures/upload"))
             .respond_with(move |req: &Request| {
-                #[derive(Debug, serde::Deserialize)]
+                #[derive(Debug, serde::Deserialize, Serialize)]
                 #[serde(transparent)]
                 struct Parameters(BTreeMap<OwnedUserId, SignedKeys>);
 
                 let params: Parameters = req.body_json().unwrap();
 
-                let mut device_keys = device_keys_clone.lock().unwrap();
-                let known_devices = known_devices_clone.lock().unwrap();
-                for (user_id, signed_keys) in params.0 {
-                    for (key_id, signature) in signed_keys.iter() {
-                        if known_devices.contains(key_id) {
-                            device_keys
-                                .entry(user_id.clone())
-                                .or_default()
-                                .insert(key_id.to_owned(), signature.to_owned());
+                let mut keys = k.lock().unwrap();
+
+                for (p_user, p_signed_keys) in params.0 {
+                    for (p_key_id, p_json_diff) in p_signed_keys.iter() {
+                        // Try to find a field in keys.master.
```

Comment:
Oh my god, of course the `/keys/signatures/upload` endpoint has the same problem as interactive verification. You have to guesstimate if a key ID belongs to some cross-signing key or if it belongs to the device keys.

We had a vulnerability in some of the client SDKs because of that, since you user can control what device ID their device will get, they can create a device ID that is the same as the ID of a cross-signing key, creating a clash...